### PR TITLE
Implement initial component updater for brave-core

### PR DESCRIPTION
**Pull request by bbondy**
## Summary:

This PR implements a component update server for brave-core in Go.

The component update server uses an XML schema defined by Google.
This server is compatible with Google's component update server, so it is a drop-in replacement for their endpoint.

See also the component update server we use in muon which was written in Node here:
https://github.com/brave/vault-updater/blob/master/src/controllers/extensions.js

## Related issues

- Security review: https://github.com/brave/internal/issues/406
- Devops deploy: https://github.com/brave/devops/issues/167

## Objectives:

- Allow updating our own extensions
- For what we don't support, redirect to Google's component update server
- Allow for blacklisted extensions

## Known limitations:

- Intentionally to keep things simple I did not use any sort of a database for now.
- I might more simply just split out a package at a different repo to manage the extensions in a follow-up PR or use redis or mongo or something like that.
- In a follow up I'll change `Contains` to use some kind of a hash-set lookup instead of the O(N) search.

## Reviewer notes:

- This PR is reviewed most easily by doing 1 commit at a time as care was taken to split out the commits into logical steps.
- After this is reviewed, I will post a task to do a security review before merging.

## Run lint:

`make lint`

## Run tests:

`make test`

## Run go-update:

`make`

### CODE REVIEW

**Comment by bsclifton** [comment]
cc: @AilinLiao

File: server/server.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	ctx = lg.WithLoggerContext(ctx, logger)
+	return ctx, logger
+}
+
+func setupRouter(ctx context.Context, logger *logrus.Logger) (context.Context, *chi.Mux) {
+	r := chi.NewRouter()
+	r.Use(chiware.RequestID)
+	r.Use(chiware.RealIP)
+	r.Use(chiware.Heartbeat("/"))
+	r.Use(chiware.Timeout(60 * time.Second))
```
Comment (evq) [inline server/server.go]: I might recommend something like the `RequestLogger` middleware from `bat-go` that handles panic recovery / sending to sentry (in addition to request logging). If you want to reuse that middleware you can use the tagged bat-go release via `dep ensure -add github.com/brave-intl/bat-go@0.1.0` then just import the `middleware` package

File: extension/xml.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+		app.UpdateCheck.Manifest.Packages.Package = append(app.UpdateCheck.Manifest.Packages.Package, pkg)
+		response.Apps = append(response.Apps, app)
+	}
+	output, err := xml.MarshalIndent(response, "", "    ")
+	return output, err
+}
+
+// UnmarshalXML unmarshals the passed component update server XML data and
+// unmarshals it into an array of Extension.
+func UnmarshalXML(data []byte) (Extensions, error) {
```
[the remark this answers is not in the document]
Comment (bbondy) [reply extension/xml.go]: The suggestion for the Unmarshaller interface worked well.
I had trouble getting the Marshaller interface working though.

The branch I attempted both is here:
https://github.com/brave/go-update/commit/5a80340d38adc8374a07ceca0ac2302574c5b920

In particular `e.EncodeElement` was always converting to something like <Extension><ID>...</ID>....</Extension> even know I pass in a `Response` struct.

I think I could make it work by doing each thing individually but that'd be more work.

I think using the interface was more meant when you're actually encoding the struct you want to implement the interface for. In my case it's more of a request XML for the data in the struct, and a response XML for the data in the struct, and not the struct itself.

For symmetry I think I'd like to keep both the same, but let me know if you see something I'm doing wrong in the branch above.

File: server/server.go (hunk 1 above)
Comment (bbondy) [reply server/server.go]: This and things other than the XML one below are implemented and pushed out.

**Comment by bbondy** [comment]
OK all comments addressed and PR is updated. Thanks for the help on Slack disambiguating the slice vs the type which I was just thinking of as an alias in my head.  Updated to other suggestions like using a receiver for the now `Contains` and `FilterForUpdates` methods.

File: controller/controller.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+		http.Error(w, fmt.Sprintf("Error reading body %v", err), http.StatusBadRequest)
+		return
+	}
+
+	// Special case, if there's only 1 extension in the request and it is not something
+	// we know about, redirect the client to google component update server.
+	if len(extensions) == 1 {
+		_, err := allExtensions.Contains(extensions[0].ID)
+		if err != nil {
+			http.Redirect(w, r, "http://update.googleapis.com/service/update2?braveRedirect=true", http.StatusTemporaryRedirect)
```
Comment (diracdeltas) [inline controller/controller.go]: just curious, is there a reason for the braveRedirect param?

Comment (diracdeltas) [reply controller/controller.go]: also this should be HTTPS not HTTP

File: extension/data.go [hunk 4]
```diff
@@ -0,0 +1,301 @@
+package extension
+
+// OfferedExtensions contains the extensions our server handles
+// ourselves without redirecting to Google's component updater.
+var OfferedExtensions = Extensions{
+	{
+		ID:          "ldimlcelhnjgpjjemdjokpgeeikdinbm",
```
Comment (diracdeltas) [inline extension/data.go]: i assume we just threw away the private keys that these IDs correspond to?

**Comment by bbondy** [comment]
http was just for testing and accidentally pushed it in the last commit, reverted that back to https.

File: extension/data.go [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+		Version:     "1.0.1",
+		SHA256:      "992a7a2147cc1593c638de7c1e80903a0f77e24088ee105f8c841b325b957519",
+		Title:       "Brave Ad Block Updater (BGR: Bulgarian Adblock list)",
+		URL:         "",
+		Blacklisted: false,
+	}, {
+		ID:          "oofiananboodjbbmdelgdommihjbkfag",
+		Version:     "1.0.1",
+		SHA256:      "c3c40af8bff5aacf90cb4176b820644756665de82c5e37170c7ad6f2878fea36",
+		Title:       "Brave HTTPS Everywhere Updater",
```
[the remark this answers is not in the document]
Comment (bbondy) [reply extension/data.go]: Per discussion before we do this I'd like for us to see if the speed gains are negligible.  Alternately if they are not negligible we could invest into tooling to keep it up to date.

File: extension/data.go (hunk 4 above)
Comment (bbondy) [reply extension/data.go]: They're in 1pwd.

File: README.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+`make lint`
+
+## Run tests:
+
+`make test`
+
+## Run go-update:
+
+`make`
```
[the remark this answers is not in the document]
Comment (bbondy) [reply README.md]: I don't think there's a formal spec but I'll find something that briefly describes it. It's mostly figured out by using wireshark.  Example requests and responses are in the tests. 

I'll improve the README, thanks for the suggestion.

File: controller/controller.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+func UpdateExtensions(w http.ResponseWriter, r *http.Request) {
+	log := lg.Log(r.Context())
+	defer func() {
+		err := r.Body.Close()
+		if err != nil {
+			log.Errorf("Error closing body stream: %v", err)
+		}
+	}()
+
+	body, err := ioutil.ReadAll(r.Body)
```
[the remark this answers is not in the document]
Comment (bbondy) [reply controller/controller.go]: I'll look into it.

File: extension/data.go [hunk 8]
```diff
@@ -0,0 +1,301 @@
+package extension
+
+// OfferedExtensions contains the extensions our server handles
+// ourselves without redirecting to Google's component updater.
+var OfferedExtensions = Extensions{
```
[the remark this answers is not in the document]
Comment (bbondy) [reply extension/data.go]: Yep, see the PR description which describes this initial limitation. 
There will be follow up work that moves this somewhere more manageable.

File: extension/extension.go [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+func (extensions *Extensions) FilterForUpdates(extensionsToCheck Extensions) Extensions {
+	filteredExtensions := Extensions{}
+	for _, extensionBeingChecked := range extensionsToCheck {
+		foundExtension, err := extensions.Contains(extensionBeingChecked.ID)
+		if err == nil {
+			if !foundExtension.Blacklisted && CompareVersions(extensionBeingChecked.Version, foundExtension.Version) < 0 {
+				filteredExtensions = append(filteredExtensions, foundExtension)
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (riastradh-brave) [reply extension/extension.go]: Sorry, missed that you had already mentioned in the PR description that this will be fixed in a subsequent revision.

**Comment by bbondy** [comment]
> Generally looks OK except for some possible denial of service vectors that should be addressed. Would like a statement of what it does for whom and with what resources in the README, though, so I don't have to ask or figure it out by reading code.

It's a bit of a stretch to say what it does and for whom needs to be figured out by reading code since there's a comprehensive PR description. But yes I'll improve the README.

File: extension/extension.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (bbondy) [reply extension/extension.go]: Yep plan to change as indicated in the PR description, but also of note `updateRequest` is always len 1 for our usage.

**Comment by bbondy** [comment]
@riastradh-brave @evq new commits were added:
https://github.com/brave/go-update/pull/1/commits/8c0822d5f272f550aa0af9288cf243371a205b60
https://github.com/brave/go-update/pull/1/commits/0058994c05e27af356547505b6aea2c04b8b1b7e
https://github.com/brave/go-update/pull/1/commits/b2b66038d155960258af0843c38af6b0733c33fe

Thank you for reviewing.

Could you let me know if there are any other concerns? We'd ideally like to have this launched for use in brave-core this week. Work will continue on it incrementally over time, this is just meant to be the initial commits.

File: server/server_test.go [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+	testCall(t, server, http.MethodPost, "", requestBody, http.StatusOK, expectedResponse)
+
+	// Unkonwn extension ID goes to Google server
+	requestBody = extensiontest.ExtensionRequestFnFor("aaaaaaaaaaaaaaaaaaaa")("0.0.0")
+	expectedResponse = ""
+	testCall(t, server, http.MethodPost, "", requestBody, http.StatusTemporaryRedirect, expectedResponse)
+
+	// Make sure a huge request body does not crash the server
+	data := make([]byte, 1024 * 1024 * 11) // 11 MiB
+	rand.Read(data)
```
Comment (evq) [inline server/server_test.go]: @bbondy CI is failing due to a few lint issues. For one you will need to run `gofmt -s` on this file, for the other you can add `// #nosec` on the end of this line to disable the gas (gosec) linter

**Comment by bbondy** [comment]
Oops, fixed lint and re-pushed @evq . Simply used `cyrpto/rand` to avoid the nosec.

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (riastradh-brave) [reply README.md]: Thanks, the additions are very helpful!

File: extension/data.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (riastradh-brave) [reply extension/data.go]: Got it, thanks!

File: extension/extension.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (riastradh-brave) [reply extension/extension.go]: Is that guaranteed by the requests sent by the client we have written, or is that guaranteed by something in this go-update server?  The latter is what's interesting for resilience to DoS.

[the remark this answers is not in the document]
Comment (bbondy) [reply extension/extension.go]: As noted for our usage.
If you want to dig deeper we have 42 extensions. So it is `O(42N) = O(N)`.
But this is something I already specifically called out as a task I will do in an upcoming PR. 
I don't typically code 100% of everything in a single PR.

**Pull request merged**